core/src/main/java/org/eclipse/hono/util/Futures.java

```diff
@@ -111,6 +113,51 @@ public static <T> Future<T> executeBlocking(final Vertx vertx, final BlockingCod
         return result.future();
     }
 
+    /**
+     * Executes some code, ensuring it is run on the given context or a duplicated context of the given context.
+     * <p>
+     * If invoked from a context that is either equal to the given context or a duplicated context of it, the code is
+     * run directly. Otherwise it is scheduled to be run asynchronously on the given context.
+     *
+     * @param <T> The type of the result that the code produces.
+     * @param requiredContextOrContextRoot The context to run the code on.
+     * @param codeToRun The code to execute. The code is required to either complete or
+     *                  fail the promise that is passed into the handler.
+     * @return The future containing the result of the promise passed in to the handler for
+     *         executing the code. The future thus indicates the outcome of executing
+     *         the code. The future will always be failed if the required context is {@code null}.
+     * @throws NullPointerException If codeToRun is {@code null}.
+     */
+    public static <T> Future<T> executeOnContextWithSameRoot(
+            final Context requiredContextOrContextRoot,
+            final Handler<Promise<T>> codeToRun) {
+
+        Objects.requireNonNull(codeToRun);
+
+        final Promise<T> result = Promise.promise();
+        if (requiredContextOrContextRoot == null) {
+            result.fail(new IllegalStateException("no context to run on"));
```

Comment:
shouldn't this be a NPE instead? Also, IMHO we do not need to create the Promise in this case but instead can simply add a `Objects.requireNonNull(requiredContextorContextRoot)` ...